File: frontend/src/services/api/pricing.ts
/**
 * Pricing API - 模型价格与 USD 汇率
 *
 * 汇率表由组件按需拉取，模块内缓存 30 分钟（后端本身 24h 同步一次）；
 * 拉取失败返回 null，调用方回落 USD 展示。
 */

import { API_BASE } from "./config";
import { authFetch } from "./fetch";

export interface FxRatesResponse {
  base: string;
  rates: Record<string, number>;
  synced_at: string | null;
}

export interface PricingLookupResponse {
  found: boolean;
  source: string;
  rates: Record<string, number | null> | null;
  matched_provider: string;
  matched_model_id: string;
}

export interface PricingStatusResponse {
  prices: { entry_count: number; source_url: string; synced_at: string | null };
  fx: { base: string; rate_count: number; synced_at: string | null };
}

export interface PricingSyncResponse extends PricingStatusResponse {
  refreshed: boolean;
  error: string | null;
}

export interface PricingBackfillResponse {
  scanned: number;
  priced: number;
  still_unpriced: number;
  unpriced_models: Record<string, number>;
  dry_run: boolean;
}

const FX_CACHE_TTL_MS = 30 * 60 * 1000;

let fxCache: { doc: FxRatesResponse | null; at: number } | null = null;
let fxInflight: Promise<FxRatesResponse | null> | null = null;

/** 获取 USD 基准汇率表（带缓存；失败返回 null） */
export async function getFxRates(force = false): Promise<FxRatesResponse | null> {
  const now = Date.now();
  if (!force && fxCache && now - fxCache.at < FX_CACHE_TTL_MS) {
    return fxCache.doc;
  }
  if (!force && fxInflight) return fxInflight;

  fxInflight = authFetch<FxRatesResponse>(`${API_BASE}/api/pricing/rates`)
    .then((doc) => {
      fxCache = { doc, at: Date.now() };
      return doc;
    })
    .catch(() => {
      // 失败保留旧缓存（若有），调用方回落 USD
      if (fxCache) fxCache = { doc: fxCache.doc, at: Date.now() };
      return fxCache?.doc ?? null;
    })
    .finally(() => {
      fxInflight = null;
    });
  return fxInflight;
}

export const pricingApi = {
  /** 管理员：手动同步 models.dev 价格 + 汇率 */
  async sync(): Promise<PricingSyncResponse> {
    const doc = await authFetch<PricingSyncResponse>(`${API_BASE}/api/pricing/sync`, {
      method: "POST",
    });
    void getFxRates(true).catch(() => null);
    return doc;
  },

  /** 管理员：补算存量 usage_logs 费用（幂等） */
  async backfillUsage(dryRun = false): Promise<PricingBackfillResponse> {
    return authFetch<PricingBackfillResponse>(
      `${API_BASE}/api/pricing/backfill-usage?dry_run=${dryRun}`,
      { method: "POST" },
    );
  },

  /** 管理员：同步状态 */
  async status(): Promise<PricingStatusResponse> {
    return authFetch<PricingStatusResponse>(`${API_BASE}/api/pricing/status`);
  },

  /** 管理员：按模型标识查询单价 */
  async lookup(params: {
    value: string;
    provider?: string;
    model_id?: string;
  }): Promise<PricingLookupResponse> {
    const searchParams = new URLSearchParams({ value: params.value });
    if (params.provider) searchParams.append("provider", params.provider);
    if (params.model_id) searchParams.append("model_id", params.model_id);
    return authFetch<PricingLookupResponse>(
      `${API_BASE}/api/pricing/lookup?${searchParams.toString()}`,
    );
  },
};

File: src/kernel/schemas/pricing.py
"""Pricing schemas: 汇率查询、同步状态、模型价格查询。"""

from datetime import datetime
from typing import Optional

from pydantic import BaseModel, Field


class FxRatesResponse(BaseModel):
    """USD 基准汇率表（前端本地货币换算用）。"""

    base: str = "USD"
    rates: dict[str, float] = Field(default_factory=dict)
    synced_at: Optional[datetime] = None


class PricingSnapshotInfo(BaseModel):
    """models.dev 价格快照状态。"""

    entry_count: int = 0
    source_url: str = ""
    synced_at: Optional[datetime] = None


class PricingFxInfo(BaseModel):
    """汇率表状态。"""

    base: str = "USD"
    rate_count: int = 0
    synced_at: Optional[datetime] = None


class PricingStatusResponse(BaseModel):
    """同步状态（管理员）。"""

    prices: PricingSnapshotInfo = Field(default_factory=PricingSnapshotInfo)
    fx: PricingFxInfo = Field(default_factory=PricingFxInfo)


class PricingSyncResponse(PricingStatusResponse):
    """手动同步结果（管理员）。"""

    refreshed: bool = False
    error: Optional[str] = None


class PricingLookupResponse(BaseModel):
    """按模型标识查询解析后的单价（管理员，模型表单提示用）。"""

    found: bool = False
    source: str = ""
    rates: Optional[dict[str, Optional[float]]] = None
    matched_provider: str = ""
    matched_model_id: str = ""


class PricingBackfillResponse(BaseModel):
    """存量 usage_logs 费用回填结果。"""

    scanned: int = 0
    priced: int = 0
    still_unpriced: int = 0
    unpriced_models: dict[str, int] = Field(default_factory=dict)
    dry_run: bool = False
    lock_contended: bool = False
